cli_tools/common/imagefile/qemu_img.go

```diff
@@ -53,34 +57,112 @@ func NewInfoClient() InfoClient {
 	if !files.Exists(filename) {
-		return info, fmt.Errorf("file %q not found", filename)
+		err = fmt.Errorf("file %q not found", filename)
+		return
+	}
+
+	jsonTemplate, err := client.getFileInfo(ctx, filename)
+	if err != nil {
+		err = daisy.Errf("Failed to inspect file %v: %v", filename, err)
+		return
+	}
+	info.Format = lookupFileFormat(jsonTemplate.Format)
+	info.ActualSizeBytes = jsonTemplate.ActualSizeBytes
+	info.VirtualSizeBytes = jsonTemplate.VirtualSizeBytes
+
+	checksum, err := client.getFileChecksum(ctx, filename, info.VirtualSizeBytes)
+	if err != nil {
+		err = daisy.Errf("Failed to calculate file '%v' checksum by qemu: %v", filename, err)
+		return
 	}
+
+	info.Checksum = checksum
+	return
+}
+
+func (client defaultInfoClient) getFileInfo(ctx context.Context, filename string) (*fileInfoJsonTemplate, error) {
 	cmd := exec.CommandContext(ctx, "qemu-img", "info", "--output=json", filename)
 	out, err := cmd.Output()
+	err = constructCmdErr(string(out), err, "inspection failure")
 	if err != nil {
-		var exitError *exec.ExitError
-		if errors.As(err, &exitError) {
-			return info, fmt.Errorf("inspection failure: %w, stderr: %s", err,
-				exitError.Stderr)
-		}
-		return info, fmt.Errorf("inspection failure: %w", err)
+		return nil, err
 	}
-	jsonTemplate := struct {
-		Filename         string `json:"filename"`
-		Format           string `json:"format"`
-		ActualSizeBytes  int64  `json:"actual-size"`
-		VirtualSizeBytes int64  `json:"virtual-size"`
-	}{}
+
+	jsonTemplate := fileInfoJsonTemplate{}
 	if err = json.Unmarshal(out, &jsonTemplate); err != nil {
-		return info, fmt.Errorf("failed to inspect %q: %w", filename, err)
+		return nil, daisy.Errf("failed to inspect %q: %w", filename, err)
+	}
+	return &jsonTemplate, err
+}
+
+func (client defaultInfoClient) getFileChecksum(ctx context.Context, filename string, virtualSizeBytes int64) (checksum string, err error) {
+	// Check size = 200000*512 = 100MB
+	checkCount := int64(200000)
+	blockSize := int64(512)
+	blockCount := virtualSizeBytes / blockSize
+	skips := []int64{0, int64(2000000) - checkCount, int64(20000000) - checkCount, blockCount - checkCount}
+	tmpOutFilePrefix := "out" + pathutils.RandString(5)
+	for i, skip := range skips {
+		tmpOutFileName := fmt.Sprintf("%v%v", tmpOutFilePrefix, i)
+		defer os.Remove(tmpOutFileName)
+
+		if skip < 0 {
+			skip = 0
+		}
+
+		// Write 100MB data to a file.
+		cmd := exec.CommandContext(ctx, "qemu-img", "dd", fmt.Sprintf("if=%v", filename),
+			fmt.Sprintf("of=%v", tmpOutFileName), fmt.Sprintf("bs=%v", blockSize),
+			fmt.Sprintf("count=%v", skip+checkCount), fmt.Sprintf("skip=%v", skip))
+		var out []byte
+		out, err = cmd.Output()
+		err = constructCmdErr(string(out), err, "inspection for checksum failure")
+		if err != nil {
+			return
+		}
+
+		// Calculate checksum for the 100MB file.
+		cmd = exec.CommandContext(ctx, "md5sum", tmpOutFileName)
```

Comment:
In case unseen, the Go standard library allows you to calculate MD5s. This might be helpful to avoid having to parse the output of `md5sum`, and making sure that `md5sum` is available at runtime

https://pkg.go.dev/crypto/md5